bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/Powerline546EHandler.java

```diff
@@ -74,41 +81,123 @@ public Powerline546EHandler(Bridge bridge, HttpClient httpClient,
+                    new QuantityType<>(device.getPowermeter().getPower(), SmartHomeUnits.WATT));
+            updateThingChannelState(CHANNEL_VOLTAGE,
+                    new QuantityType<>(device.getPowermeter().getVoltage(), SmartHomeUnits.VOLT));
+        }
+        if (device.isSwitchableOutlet() && device.getSwitch() != null) {
+            updateThingChannelState(CHANNEL_MODE, new StringType(device.getSwitch().getMode()));
+            updateThingChannelState(CHANNEL_LOCKED,
+                    BigDecimal.ZERO.equals(device.getSwitch().getLock()) ? OpenClosedType.OPEN : OpenClosedType.CLOSED);
+            updateThingChannelState(CHANNEL_DEVICE_LOCKED,
+                    BigDecimal.ZERO.equals(device.getSwitch().getDevicelock()) ? OpenClosedType.OPEN
+                            : OpenClosedType.CLOSED);
+            if (device.getSwitch().getState() == null) {
+                updateThingChannelState(CHANNEL_OUTLET, UnDefType.UNDEF);
+            } else {
+                updateThingChannelState(CHANNEL_OUTLET,
+                        SwitchModel.ON.equals(device.getSwitch().getState()) ? OnOffType.ON : OnOffType.OFF);
+            }
+        }
+    }
+
+    /**
+     * Updates thing properties.
+     *
+     * @param device the {@link AVMFritzBaseModel}
+     */
+    private void updateProperties(AVMFritzBaseModel device) {
+        Map<String, String> editProperties = editProperties();
+        editProperties.put(Thing.PROPERTY_FIRMWARE_VERSION, device.getFirmwareVersion());
+        updateProperties(editProperties);
+    }
+
+    /**
+     * Updates thing channels.
+     *
+     * @param channelId ID of the channel to be updated.
+     * @param state State to be set.
+     */
+    private void updateThingChannelState(String channelId, State state) {
+        Channel channel = thing.getChannel(channelId);
+        if (channel != null) {
+            updateState(channel.getUID(), state);
+        } else {
+            logger.debug("Channel '{}' in thing '{}' does not exist, recreating thing.", channelId, thing.getUID());
+            createChannel(channelId);
+        }
+    }
+
+    /**
+     * Creates new channels for the thing.
+     *
+     * @param channelId ID of the channel to be created.
+     */
+    private void createChannel(String channelId) {
+        ThingHandlerCallback callback = getCallback();
+        if (callback != null) {
+            ChannelUID channelUID = new ChannelUID(thing.getUID(), channelId);
+            ChannelTypeUID channelTypeUID = CHANNEL_BATTERY.equals(channelId)
+                    ? new ChannelTypeUID("system:battery-level")
+                    : new ChannelTypeUID(BINDING_ID, channelId);
+            Channel channel = callback.createChannelBuilder(channelUID, channelTypeUID).build();
+            updateThing(editThing().withoutChannel(channelUID).withChannel(channel).build());
         }
-        super.updateThingFromDevice(thing, device);
+    }
```

Comment:
This logic is also used in AVMFritzBaseThingHandler so it should be refactored into a utility class.